### PR TITLE
Feat/wcfm to dokan

### PR DESCRIPTION
**Pull request by Aunshon**

### CODE REVIEW

File: dokan-migrator.php
Comment (tanmayjay) [inline dokan-migrator.php]: Don't apply this here. It should be applied just before the beginning of the migration and removed just after the process is finished.

File: dokan-migrator.php
Comment (tanmayjay) [inline dokan-migrator.php]: Can we add this to new dokan admin notices?

File: includes/Ajax.php
Comment (tanmayjay) [inline includes/Ajax.php]: Here you don't need the `$_REQUEST` as parameter. You can directly access the global parameter as below:
```
        if ( ! isset( $_REQUEST['nonce'] ) || ! wp_verify_nonce( sanitize_key( wp_unslash( $_REQUEST['nonce'] ) ), 'dokan_migrator_nonce' ) ) {
            wp_send_json_error( array( 'message' => __( 'Nonce verification failed!', 'dokan-migrator' ) ) );
        }
        ```

File: includes/Ajax.php
Comment (tanmayjay) [inline includes/Ajax.php]: Instead of using `// phpcs:ignore`, make the declaration more specific mentioning which rule you're ignoring. For example, `// phpcs:ignore Squiz.Arrays.ArrayDeclaration.SingleLineNotAllowed`

File: includes/Ajax.php
Comment (tanmayjay) [inline includes/Ajax.php]: Anywhere it has been done this way, must be removed. No need to `unslash` the whole `$_REQUEST` and store it to another variable. Just do the process as follows:
```
$import = ! empty( $_POST['import'] ) ? sanitize_text_field( wp_unslash( $_POST['import'] ) ) : 'vendor';
$migratable = ! empty( $_POST['migratable'] ) ? boolval( wp_unslash( $_POST['migratable'] ) ) : false;
```

Comment (tanmayjay) [reply includes/Ajax.php]: Also, why are the posted values parsed using `$_REQUEST`?
Try to avoid this if not necessary. If not the process is taking any `$_GET` variable, then `$_POST` is appropriate for `Ajax` calls.

File: dokan-migrator.php
Comment (tanmayjay) [inline dokan-migrator.php]: This method could be renamed as 'hooks' so that all necessary hooks for now and for future could be initiated inside the method.

File: includes/Abstracts/OrderMigration.php
Comment (tanmayjay) [inline includes/Abstracts/OrderMigration.php]: Use `dokan_current_datetime()` instead of `date()` or `gmdate()`

File: includes/Admin/Menu.php
Comment (tanmayjay) [inline includes/Admin/Menu.php]: Wouldn't it be better if this menu was registered under the `Dokan` menu instead of a dedicated menu?

File: includes/Abstracts/OrderMigration.php
Comment (tanmayjay) [inline includes/Abstracts/OrderMigration.php:197]: Can we use `dokan_sync_insert_order()` function instead of repeating the process in a new function?

**Review by tanmayjay — CHANGES_REQUESTED** [review]

File: templates/template-active-vendor-dashboard.php
Comment (tanmayjay) [inline templates/template-active-vendor-dashboard.php]: The variable has to be escaped while printing. So, write it in the following way:
```

                    printf(
                        /* translators: 1$s: opening anchor tag, 2$s: closing anchor tag */
                        esc_html__( '%1$sActive now%2$s', 'dokan-migrator' ),
                        '<a id="dokan-migrator-active-v-dash" href="#">',
                        '</a>'
                    );
```

File: templates/template-active-vendor-dashboard.php
Comment (tanmayjay) [inline templates/template-active-vendor-dashboard.php:30]: Can this portion of script be moved to the main script so that it can be executed from there?

File: src/components/App.js
Comment (tanmayjay) [inline src/components/App.js]: Don't use plain text. Kindly check all the plain texts and make them translatable using `i18n`

**Review by tanmayjay — CHANGES_REQUESTED** [review]

File: includes/Abstracts/OrderMigration.php
Comment (tanmayjay) [inline includes/Abstracts/OrderMigration.php]: This method is redundant. The default `dokan()->order->create_sub_order()` could be used to do this.
Follow the below steps and also check if the given steps satisfy the needs.
```
    /**
     * Creates sub orders for each seller.
     *
     * @since 1.0.0
     *
     * @param int $seller_id
     * @param array $seller_products
     *
     * @return \WC_Order
     */
    public function create_sub_order( $seller_id, $seller_products ) {
        /*
         * Before creating sub order, we need to convert the meta key according to
         * Dokan meta key for order items as those data will be parsed while creating
         * sub orders.
         */
        $this->map_shipping_method_item_meta();

        return dokan()->order->create_sub_order( $this->order, $seller_id, $seller_products );
    }

    /**
     * Converts the order item meta key according to
     * Dokan meta key as those data will be parsed
     * while creating sub orders.
     *
     * @since 1.0.0
     *
     * @return void
     */
    public function map_shipping_method_item_meta() {
        if ( ! $this->order instanceof \WC_Order ) {
            return;
        }

        $shipping_methods = $this->order->get_shipping_methods();
        if ( empty( $shipping_methods ) ) {
            return;
        }

        foreach ( $shipping_methods as $method_item_id => $shipping_object ) {
            wc_update_order_item_meta(
                $method_item_id,
                'seller_id',
                wc_get_order_item_meta( $method_item_id, 'vendor_id', true )
            );
        }
    }
```
And if it works, the `create_shipping` method inside this class will not be needed.

**Review by tanmayjay — CHANGES_REQUESTED** [review]

File: includes/Abstracts/OrderMigration.php
Comment (tanmayjay) [reply includes/Abstracts/OrderMigration.php]:
```
dokan_current_datetime()->modify( $created_date )->format( 'Y-m-d H:i:s' );
```

File: includes/Abstracts/OrderMigration.php
Comment (tanmayjay) [inline includes/Abstracts/OrderMigration.php]: Using so many parameters for a method is not a good practice. Kindly reduce the number of parameters by passing them inside an array. For example, the parameters can be `$order_id, $seller_id, array $data[ rest of the parameters go inside the array ]`

File: includes/Integrations/Wcfm/OrderMigrator.php
Comment (tanmayjay) [inline includes/Integrations/Wcfm/OrderMigrator.php]: Instead of using `json_encode`, kindly use `wp_json_encode` as recommended.

File: includes/Integrations/Wcfm/OrderMigrator.php
Comment (tanmayjay) [inline includes/Integrations/Wcfm/OrderMigrator.php]: Instead of using `round`, you could use `Automattic\WooCommerce\Utilities\NumberUtil::round` as recommended.

File: includes/Integrations/Wcfm/OrderMigrator.php
Comment (tanmayjay) [inline includes/Integrations/Wcfm/OrderMigrator.php]: This declaration is not needed.

File: includes/Integrations/Wcfm/OrderMigrator.php
Comment (tanmayjay) [inline includes/Integrations/Wcfm/OrderMigrator.php]: Rewrite these two lines as follows:
```
                if ( $c_refunded_qty && $line_item->get_quantity() === $c_refunded_qty ) {
                    $is_partially_refunded = false;
                }
                $is_refunded = ! $is_partially_refunded;
```

File: includes/Integrations/Wcfm/OrderMigrator.php
Comment (tanmayjay) [inline includes/Integrations/Wcfm/OrderMigrator.php]: Why this condition is necessary? This block is already inside the same condition. I guess it should be like below:
```
                    if ( ! empty( $refund_tax ) && is_array( $refund_tax ) ) {
                        if ( isset( $refund_tax['total'] ) ) {
                            $refund_tax_total = $refund_tax['total'];
                        } else {
                            foreach ( $refund_tax as $refund_tax_id => $refund_tax_price ) {
                                $refund_tax_total += (float) $refund_tax_price;
                            }
                        }
                    }
```
Kindly verify.

File: includes/Integrations/Wcfm/OrderMigrator.php
Comment (tanmayjay) [inline includes/Integrations/Wcfm/OrderMigrator.php]: This shorthand is not a good practice. Expand this kind of shorthand.
```
if ( ! $shipping_vendor_id ) {
    $shipping_vendor_id = 0;
}
```
Also, this line should not be needed. I guess doing the following in the previous line will do the trick.
```
shipping_vendor_id = (int) $order_item_shipping->get_meta( 'vendor_id', true );
```

File: includes/Integrations/Wcfm/OrderMigrator.php
Comment (tanmayjay) [inline includes/Integrations/Wcfm/OrderMigrator.php]: This so many conditions is not needed. Just do the following:
```
if ( ! empty( $vendor_shipping[ $vendor_id ]['shipping_item_id'] ) )
```

File: includes/Integrations/Wcfm/OrderMigrator.php
Comment (tanmayjay) [inline includes/Integrations/Wcfm/OrderMigrator.php:285]: Kindly add some comments on each block describing why you are doing that.

**Review by tanmayjay — CHANGES_REQUESTED** [review]

**Comment by shashwatahalder01** [comment]
Scenario:
admin commission shouldn't include shipping & tax after running the migrator

Steps to reproduce:
1. create multiple order on wcfm
2. migrate to dokan
3. see admin report > all logs > admin commission

Screenshot:
dc1
dc2
dc3

**Comment by shashwatahalder01** [comment]
Scenario:
-  [x] Don't migrate some vendor info:
         vendor first name and last name (account details)
         phone num (store details)

**Comment by shashwatahalder01** [comment]
Scenario:
-  [x] Error on several places for migrated orders with refunds(for both partial & full)
    Erroneous places:
-  [x] Vendor dashboard (sales, earning,..), Vendor order table, vendor reports, vendor statements, vendor balance, admin dashboard (commission, net sales), admin reports(all logs), 

Steps to reproduce:
1. place multiple orders on WCFM
2. refund(both partial and full) some of the orders
3. migrate to dokan
4. go to mentioned  places

Screenshot:
 Full refund
wc2 0
wc2 1
partial refund
wc3
wc4

wc5
wc6
wc7
wc8
wc9

**Comment by shashwatahalder01** [comment]
Scenario:
withdraw migration issues:
  -  [x] admin can't approve any pending withdraw request (migrated)
  -  [x] missing withdraw method on vendor withdraw menu ( exists on admin withdraw table)
  -  [x] multiple pending withdraw request on vendor withdraw menu (only last one should be on vendor withdraw menu)
  
Steps:
1. place multiple order and withdraw request
2. run migratory
3. go to admin withdraw menu & vendor withdraw menu

Screenshot:

wc12
wc13
wc15
wc17

**Comment by shashwatahalder01** [comment]
Scenario:
withdraw settings aren't migrated like:
- minimum withdrawal amount
- Order Status for Withdraw 
- withdraw threshold
- disbursement settings

steps to reproduce:
1. set withdraw setting on wcfm
2. run the migrator
3. see dokan withdraw settings

Screenshot:
wc20

wc22

**Comment by shashwatahalder01** [comment]
Scenario:
bank details are migrated but payment method isn't connected as account type is missing on wcfm.
Set account type as personal  by default for migrated vendors with bank transfer payment method.


Steps to reproduce:
1. set bank details for a vendor
2. run the migrator
3. see vendor payment method

Screentshot:
wc21

**Comment by shashwatahalder01** [comment]
Scenario:

wcfm rule: Requested withdraw amount is deducted from balance when a withdraw request is submitted. If the request got cancelled then requested withdraw amount is added with balance. 

dokan rule: Requested withdraw amount doesn't get deducted from balance when a withdraw request is submitted. If the request got cancelled then balance remains same. balance only updates when a request is approved.

-  [x] Migrate WCFM to Dokan with multiple pending withdraw request. Now, after migration if a pending request is cancelled by admin/vendor balance doesn't update according to dokan rule. As a result, cancelled withdraw amount is missing from vendor balance.

**Pull request merged**
